File: src/com/company/java_examples/Permutations.java
package com.company.java_examples;

import java.util.ArrayList;
import java.util.List;

public class Permutations {

    public static void main(String[] args) {
        int[] nums = {1, 2, 3};

        List<List<Integer>> result = permute(nums);
        System.out.println(result);
    }

    public static List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        fillPermutations(nums, used, new ArrayList<>(), result);
        return result;
    }

    private static void fillPermutations(int[] nums, boolean[] used, ArrayList<Integer> permutation, List<List<Integer>> result) {
        if (permutation.size() == nums.length) {
            result.add(new ArrayList<>(permutation));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) {
                continue;
            }
            used[i] = true;
            permutation.add(nums[i]);
            fillPermutations(nums, used, permutation, result);
            permutation.remove(permutation.size() - 1);
            used[i] = false;
        }
    }
}

File: src/com/company/kotlin_examples/Permutations.kt
package com.company.kotlin_examples

object Permutations {

    @JvmStatic
    fun main(args: Array<String>) {
        val nums = arrayOf(1, 2, 3)
        val result = permute(nums)
        println(result)
    }

    fun permute(nums: Array<Int>): List<List<Int>> {
        val result: MutableList<List<Int>> = ArrayList()
        val used = BooleanArray(nums.size)
        fillPermutations(nums, used, ArrayList(), result)
        return result
    }

    private fun fillPermutations(
        nums: Array<Int>,
        used: BooleanArray,
        permutation: ArrayList<Int>,
        result: MutableList<List<Int>>
    ) {
        if (permutation.size == nums.size) {
            result.add(ArrayList(permutation))
            return
        }
        for (i in nums.indices) {
            if (used[i]) {
                continue
            }
            used[i] = true
            permutation.add(nums[i])
            fillPermutations(nums, used, permutation, result)
            permutation.removeAt(permutation.size - 1)
            used[i] = false
        }
    }
}